File: cpp-implementation/src/triangle.cpp
module triangle;
import glm;
import interval;

namespace cg
{

std::optional<hitevent> get_ray_triangle_hit(const triangle& tt,
                                             std::span<const vertex> vertices,
                                             ray r, interval i,
                                             bool cull_backfaces)
{
    const vertex& v0 = vertices[tt.vertex_start];
    constexpr double epsilon = 1e-8f;

    const vertex& v1 = vertices[tt.vertex_start + 1];
    const vertex& v2 = vertices[tt.vertex_start + 2];
    const vec3& p0 = v0.p;
    const vec3& p1 = v1.p;
    const vec3& p2 = v2.p;
    const vec3& n0 = v0.n;
    const vec3& n1 = v1.n;
    const vec3& n2 = v2.n;
    // const vec3& p2 = tt.v3.p;

    using glm::mat3;
    const auto col0 = -r.dir;
    const auto col1 = p1 - p0;
    const auto col2 = p2 - p0;
    const auto col3 = r.pos - p0;

    const auto d0 = glm::determinant(mat3{col0, col1, col2});
    const interval close_to_zero{-1e-4, 1e-4};
    if (close_to_zero.contains(d0)) // paralelo al triangulo
        return std::nullopt;

    if (d0 < 0. && cull_backfaces)
        return std::nullopt;

    const auto dt = glm::determinant(mat3{col3, col1, col2});
    const double t = dt / d0;
    if (t < 0. || !i.contains(t)) // esta detras del rayo
        return std::nullopt;

    const auto du = glm::determinant(mat3{col0, col3, col2});
    const interval zero_to_one{0., 1.};
    const double u = du / d0;
    if (!zero_to_one.contains(u))
        return std::nullopt;

    const auto dv = glm::determinant(mat3{col0, col1, col3});
    const interval v_range{0., 1. - u};
    const double bary_v = dv / d0;
    if (!v_range.contains(bary_v))
        return std::nullopt;

    hitevent hit;
    hit.t = t;
    hit.p = r.at(t);
    const double bary_u = u;
    const double bary_w = 1. - bary_u - bary_v;
    const vec3 face_normal = safe_normalize(glm::cross(col1, col2));

    // compute normal
    vec3 normal =
        safe_normalize(n0 * bary_w + n1 * bary_u + n2 * bary_v, epsilon);
    if (glm::dot(normal, normal) <= epsilon)
        normal = face_normal;
    if (glm::dot(normal, r.dir) > 0.)
    {
        normal = -normal;
    }
    hit.normal = vec3{normal.x, normal.y, normal.z};

    hit.m_id = tt.material_id;
    // compute_uv
    hit.uv = v0.uv * bary_w + v1.uv * bary_u + v2.uv * bary_v;

    const vec2 duv1 = v1.uv - v0.uv;
    const vec2 duv2 = v2.uv - v0.uv;
    const double det = duv1.x * duv2.y - duv1.y * duv2.x;
    if (std::abs(det) > epsilon)
    {
        const double inv_det = 1.0 / det;
        hit.dpdu = (col1 * duv2.y - col2 * duv1.y) * inv_det;
        hit.dpdv = (col2 * duv1.x - col1 * duv2.x) * inv_det;
    }
    else
    {
        // fallback if UVs are degenerate
        vec3 a;
        if (std::abs(normal.x) > 0.9)
            a = vec3(0, 1, 0);
        else
            a = vec3(1, 0, 0);
        hit.dpdu = safe_normalize(glm::cross(normal, a), epsilon);
        hit.dpdv = safe_normalize(glm::cross(normal, hit.dpdu), epsilon);
    }
    return hit;
}

} // namespace cg

File: class-oop-implementation/src/objects/Triangle.java
package objects;

import java.util.Optional;
import math.Intersection;
import math.Interval;
import math.Matrix3;
import math.Ray;
import math.Vector2D;
import math.Vector3D;

public class Triangle implements Object3D {
    private static final float NORMAL_EPSILON = 1e-12f;
    private final Vector3D p0;
    private final Vector3D p1;
    private final Vector3D p2;
    private final Vector3D n0;
    private final Vector3D n1;
    private final Vector3D n2;
    private final Vector2D uv0;
    private final Vector2D uv1;
    private final Vector2D uv2;
    private final boolean hasVertexNormals;
    private final boolean hasVertexUvs;
    private final int materialId;

    public Triangle(Vector3D p0, Vector3D p1, Vector3D p2, int materialId) {
        this(
            p0,
            p1,
            p2,
            new Vector3D(0, 0, 0),
            new Vector3D(0, 0, 0),
            new Vector3D(0, 0, 0),
            new Vector2D(0, 0),
            new Vector2D(0, 0),
            new Vector2D(0, 0),
            false,
            false,
            materialId
        );
    }

    public Triangle(Vector3D p0, Vector3D p1, Vector3D p2, Vector3D n0, Vector3D n1, Vector3D n2, int materialId) {
        this(
            p0,
            p1,
            p2,
            n0,
            n1,
            n2,
            new Vector2D(0, 0),
            new Vector2D(0, 0),
            new Vector2D(0, 0),
            true,
            false,
            materialId
        );
    }

    public Triangle(
        Vector3D p0,
        Vector3D p1,
        Vector3D p2,
        Vector3D n0,
        Vector3D n1,
        Vector3D n2,
        Vector2D uv0,
        Vector2D uv1,
        Vector2D uv2,
        int materialId
    ) {
        this(p0, p1, p2, n0, n1, n2, uv0, uv1, uv2, true, true, materialId);
    }

    private Triangle(
        Vector3D p0,
        Vector3D p1,
        Vector3D p2,
        Vector3D n0,
        Vector3D n1,
        Vector3D n2,
        Vector2D uv0,
        Vector2D uv1,
        Vector2D uv2,
        boolean hasVertexNormals,
        boolean hasVertexUvs,
        int materialId
    ) {
        this.p0 = p0;
        this.p1 = p1;
        this.p2 = p2;
        this.n0 = n0;
        this.n1 = n1;
        this.n2 = n2;
        this.uv0 = uv0;
        this.uv1 = uv1;
        this.uv2 = uv2;
        this.hasVertexNormals = hasVertexNormals;
        this.hasVertexUvs = hasVertexUvs;
        this.materialId = materialId;
    }

    @Override
    public Optional<Intersection> isHit(Ray ray, Interval tRange) {
        // col0 = -ray.direction
        Vector3D col0 = ray.getDir().mul(-1);
        // col1 = p1 - p0
        Vector3D col1 = p1.add(p0.mul(-1));
        // col2 = p2 - p0
        Vector3D col2 = p2.add( p0.mul(-1));
        // col3 = ray.origin - p0
        Vector3D col3 = ray.getPos().add(p0.mul(-1));

        // System matrix: M = [col0 | col1 | col2]
        Matrix3 m = new Matrix3(col0, col1, col2);
        float d0 = m.determinant();

        if (Math.abs(d0) < 1e-4) {
            return Optional.empty();
        }

        float dt = new Matrix3(col3, col1, col2).determinant();
        float t = dt / d0;
        if (t < 0 || !tRange.contains(t)) {
            return Optional.empty(); // Hit is behind the ray origin
        }

        float du = new Matrix3(col0, col3, col2).determinant();
        float u = du / d0;
        if (u < 0 || u > 1) {
            return Optional.empty();
        }

        float dv = new Matrix3(col0, col1, col3).determinant();
        float v = dv / d0;
        if (v < 0 || v > (1 - u)) {
            return Optional.empty();
        }

        // Intersection confirmed
        Vector3D hitPoint = ray.at(t);
        
        final float baryU = u;
        final float baryV = v;
        final float baryW = 1f - baryU - baryV;
        Vector3D faceNormal = col1.cross(col2).normalize();
        Vector3D normal = faceNormal;
        if (hasVertexNormals) {
            normal = n0.mul(baryW).add(n1.mul(baryU)).add(n2.mul(baryV)).normalize();
            if (normal.lengthSquared() <= NORMAL_EPSILON) {
                normal = faceNormal;
            }
        }
        if (normal.dot(ray.getDir()) > 0f) {
            normal = normal.mul(-1f);
        }

        Vector2D uv = new Vector2D(0f, 0f);
        Vector3D dpdu = new Vector3D(1f, 0f, 0f);
        Vector3D dpdv = new Vector3D(0f, 1f, 0f);
        if (hasVertexUvs) {
            uv = uv0.mul(baryW).add(uv1.mul(baryU)).add(uv2.mul(baryV));
            Vector2D duv1 = uv1.sub(uv0);
            Vector2D duv2 = uv2.sub(uv0);
            float det = duv1.getX() * duv2.getY() - duv1.getY() * duv2.getX();
            if (Math.abs(det) > 1e-8f) {
                float invDet = 1f / det;
                dpdu = col1.mul(duv2.getY()).sub(col2.mul(duv1.getY())).mul(invDet);
                dpdv = col2.mul(duv1.getX()).sub(col1.mul(duv2.getX())).mul(invDet);
            } else {
                Vector3D axis = Math.abs(normal.getX()) > 0.9f ? new Vector3D(0f, 1f, 0f) : new Vector3D(1f, 0f, 0f);
                dpdu = normal.cross(axis).normalize();
                dpdv = normal.cross(dpdu).normalize();
            }
        }

        return Optional.of(new Intersection(hitPoint, normal, uv, dpdu, dpdv, t, materialId));
    }

    public Vector3D getP0() { return p0; }
    public Vector3D getP1() { return p1; }
    public Vector3D getP2() { return p2; }
    public Vector3D getN0() { return n0; }
    public Vector3D getN1() { return n1; }
    public Vector3D getN2() { return n2; }
    public Vector2D getUv0() { return uv0; }
    public Vector2D getUv1() { return uv1; }
    public Vector2D getUv2() { return uv2; }
    public boolean hasVertexNormals() { return hasVertexNormals; }
    public boolean hasVertexUvs() { return hasVertexUvs; }
    public int getMaterialId() { return materialId; }
}
